### PR TITLE
Fixed Loss Weight Adjustment in GUI: Fix Scaling and Prevent Negative Values

### PR DESCRIPTION
**Pull request by VincentHuynh1016**
### Description
I fixed the loss weight logic to ensure it does not allow negative values. I also modified the code by creating a class to override the stepBy function, enabling increments by 10 and decrements by 0.1. Additionally, I wrote tests to verify the behavior and ran the GUI to confirm everything works as expected.

### Types of changes

- [x] Bugfix
- [x] New feature
- [x] Refactor / Code style update (no logical changes)
- [ ] Build / CI changes
- [ ] Documentation Update
- [ ] Other (explain)

### Does this address any currently open issues?
https://github.com/talmolab/sleap/issues/1971

### Outside contributors checklist

- [x] Review the [guidelines for contributing](https://github.com/talmolab/sleap/blob/develop/docs/CONTRIBUTING.md) to this repository
- [x] Read and sign the [CLA](https://github.com/talmolab/sleap/blob/develop/sleap-cla.pdf) and add yourself to the [authors list](https://github.com/talmolab/sleap/blob/develop/AUTHORS)
- [x] Make sure you are making a pull request against the **develop** branch (not *main*). Also you should start *your branch* off *develop*
- [x] Add tests that prove your fix is effective or that your feature works
- [x] Add necessary documentation (if appropriate)

#### Thank you for contributing to SLEAP!
:heart:

### CODE REVIEW

File: sleap/gui/dialogs/formbuilder.py [hunk 1]
```diff
@@ -224,6 +224,28 @@ def get_results(self) -> Optional[Dict[Text, Any]]:
         return self._results
 
 
+# We are writing this method to override the spinbox in add_item so the increment/decrement is x10/x0.1
```
Comment (roomrys) [inline sleap/gui/dialogs/formbuilder.py]: Hmm, this comment looks a little long... A few things: 

1. I would suggest adding a [ruler in your vscode user settings](https://stackoverflow.com/questions/29968499/how-can-i-have-multiple-vertical-rulers-in-vs-code) at 88 characters. The linter we use ([black](https://black.readthedocs.io/en/stable/)) formats the code for us so that no lines go past 88 characters - however, it does not format/check comment-only lines so we need to enforce this ourselves).  
2. This comment looks good, but just a note: we want our codebase to follow [google-style comments and docstrings](https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings).
3. In addition to having black installed in your `sleap` conda environment, you can also install black as a [vscode extension for formatting](https://code.visualstudio.com/docs/python/formatting) and adjust either your user or workspace settings to [formatOnSave](https://code.visualstudio.com/docs/python/formatting#_format-your-code) - that way you shouldn't ever fail the github actions Lint test.

File: sleap/gui/dialogs/formbuilder.py [hunk 2]
```diff
@@ -224,6 +224,28 @@ def get_results(self) -> Optional[Dict[Text, Any]]:
...
+class ExponentialSpinBox(QtWidgets.QDoubleSpinBox):
```
Comment (roomrys) [inline sleap/gui/dialogs/formbuilder.py:227]: Add a [google-style docstring](https://google.github.io/styleguide/pyguide.html#384-classes) for this class.

File: sleap/gui/dialogs/formbuilder.py [hunk 3]
```diff
@@ -224,6 +224,28 @@ def get_results(self) -> Optional[Dict[Text, Any]]:
...
+    def stepBy(self, steps: int):
```
Comment (roomrys) [inline sleap/gui/dialogs/formbuilder.py:230]: Add a [google-style docstring](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods) for this method.

File: sleap/gui/dialogs/formbuilder.py [hunk 4]
```diff
@@ -437,13 +460,22 @@ def add_item(self, item: Dict[Text, Any]):
...
+                field = QtWidgets.QDoubleSpinBox()
+                min, max = -1000, 1000
+                field.setSingleStep(0.25)
 
-            min, max = -1000, 1000
             if "range" in item.keys():
                 min, max = list(map(float, item["range"].split(",")))
             field.setRange(min, max)
-            field.setSingleStep(0.25)
+
```
Comment (roomrys) [inline sleap/gui/dialogs/formbuilder.py]: Running `black` on this file will format this for us - either through vscode's black extension + `formatOnSave` setting or through the CLI:

```bash
black  sleap/gui/dialogs/formbuilder.py
```
.
```suggestion
```

File: sleap/gui/dialogs/formbuilder.py [hunk 5]
```diff
@@ -224,6 +224,28 @@ def get_results(self) -> Optional[Dict[Text, Any]]:
...
+        if steps == 0:
+            return
+        
+        if steps > 0:
+            newValue = currValue * 10
+            if currValue == 0:
+                newValue = 0.01
+
+        elif steps < 0:
+            newValue = currValue * ((10) ** -1)
```
Comment (roomrys) [inline sleap/gui/dialogs/formbuilder.py]: So, I believe interacting through the spinbox though it's different modes:
- buttons
- arrow keys
- mouse wheel

will pass in a different value for `steps`. I haven't tested this manually to see how if feels, but if we are multiplying/dividing by a factor of 10, then I like the current logic that only uses the +/- of `steps` to determine the output value. However, I think we should still manually test the different modes to see if we like it or not.

File: sleap/gui/dialogs/formbuilder.py [hunk 6]
```diff
@@ -224,6 +224,28 @@ def get_results(self) -> Optional[Dict[Text, Any]]:
...
+
+        # We need to keep it in the range (0 - 1000)
+        newValue = min(self.maximum(), newValue)  
+        newValue = max(self.minimum(), newValue)
+        self.setValue(newValue)
```
Comment (roomrys) [inline sleap/gui/dialogs/formbuilder.py:252]: I wonder if `setValue` implicitly enforces maximum and minimum for us? It would make sense to me if it did, but can't be too certain...

File: sleap/gui/dialogs/formbuilder.py [hunk 7]
```diff
@@ -424,6 +446,7 @@ def build_form(self, items_to_create: List[Dict[Text, Any]]):
             return
         for item in items_to_create:
             self.add_item(item)
+
```
Comment (roomrys) [inline sleap/gui/dialogs/formbuilder.py]: Running `black` on this file will format this for us - either through vscode's black extension + `formatOnSave` setting or through the CLI:

```bash
black  sleap/gui/dialogs/formbuilder.py
```
.
```suggestion
```

File: sleap/gui/dialogs/formbuilder.py [hunk 8]
```diff
@@ -437,13 +460,22 @@ def add_item(self, item: Dict[Text, Any]):
 
         # double: show spinbox (number w/ up/down controls)
         elif item["type"] == "double":
-            field = QtWidgets.QDoubleSpinBox()
+            
+            #If the label of the item is Loss Weight then we will set the min and max to 0 and 1000
```
Comment (roomrys) [inline sleap/gui/dialogs/formbuilder.py]: [Google style block comments](https://google.github.io/styleguide/pyguide.html#385-block-and-inline-comments) (also looks a bit long for a single line).

**Review by roomrys — CHANGES_REQUESTED** [review]
Logically seems fine. A few formatting suggestions and call for manual (and unit) tests.

File: .gitignore [hunk 9]
```diff
@@ -117,3 +117,4 @@ venv.bak/
 
 # OS generated files
 .DS_Store
+labels.v001.slp
```
Comment (roomrys) [inline .gitignore]: We don't want this in our PR. I am going to assume you used `git add .` and this file was included by accident? `git add <the/files/you/changed>` is always a safer option, but lately, I find myself using the [VSCode source control](https://code.visualstudio.com/docs/sourcecontrol/overview) GUI and would recommend!
```suggestion
```

File: sleap/gui/dialogs/formbuilder.py [hunk 10]
```diff
@@ -224,25 +225,32 @@ def get_results(self) -> Optional[Dict[Text, Any]]:
...
         elif steps < 0:
-            newValue = currValue * ((10) ** -1)
+            newValue = currValue * 0.1
 
-        # We need to keep it in the range (0 - 1000)
-        newValue = min(self.maximum(), newValue)  
+        else:
+            return
+
+        """Ensures value is in the range (0 - 1000)"""
```
Comment (roomrys) [inline sleap/gui/dialogs/formbuilder.py]: Triple quotes should only be used in multi-line comments (and ideally ONLY in docstrings).
```suggestion
        # Ensure value is in the range (0 - 1000)
```

File: sleap/gui/dialogs/formbuilder.py [hunk 11]
```diff
@@ -29,6 +29,7 @@
 
 from typing import Any, Dict, List, Optional, Text
 
+
```
Comment (roomrys) [inline sleap/gui/dialogs/formbuilder.py]: Lint
```suggestion
```

File: tests/gui/test_formbuilder.py [hunk 12]
```diff
@@ -132,3 +140,93 @@ def test_string_list_widget(qtbot):
...
+
+
+def test_exponential_spin_box(qtbot):
+    """I am testing the exponential spin box class itself not with the gui"""
+    widget = formbuilder.ExponentialSpinBox()
+    qtbot.addWidget(widget)
+
+    widget.setValue(1.0)
+    print(widget.value())
+    assert math.isclose(widget.value(), 1.0, rel_tol=1e-3)
```
Comment (roomrys) [inline tests/gui/test_formbuilder.py:151]: Did `==` evaluate to `False` here? Is that why we need `math.isclose`?

File: tests/gui/test_formbuilder.py [hunk 13]
```diff
@@ -132,3 +140,93 @@ def test_string_list_widget(qtbot):
...
     assert widget.text() == "zip cab"
+
+
+def test_exponential_spin_box(qtbot):
+    """I am testing the exponential spin box class itself not with the gui"""
+    widget = formbuilder.ExponentialSpinBox()
+    qtbot.addWidget(widget)
+
+    widget.setValue(1.0)
+    print(widget.value())
```
Comment (roomrys) [inline tests/gui/test_formbuilder.py]: Remove this `print` and all others.

File: tests/gui/test_formbuilder.py [hunk 14]
```diff
@@ -132,3 +140,93 @@ def test_string_list_widget(qtbot):
 
     widget.setValue(["zip", "cab"])
     assert widget.text() == "zip cab"
+
+
+def test_exponential_spin_box(qtbot):
+    """I am testing the exponential spin box class itself not with the gui"""
```
Comment (roomrys) [inline tests/gui/test_formbuilder.py]: We aren't concerned with what YOU are doing, but rather what the function is doing:
```suggestion
    """Test ExponentialSpinBox setValue and stepBy methods."""
```

> The docstring may be descriptive-style ("""Fetches rows from a Bigtable.""") or imperative-style ("""Fetch rows from a Bigtable."""), but the style should be consistent within a file.

Ref: [google-style docstrings](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods)

File: sleap/gui/dialogs/formbuilder.py [hunk 15]
```diff
@@ -224,25 +225,32 @@ def get_results(self) -> Optional[Dict[Text, Any]]:
         return self._results
 
 
-# We are writing this method to override the spinbox in add_item so the increment/decrement is x10/x0.1
 class ExponentialSpinBox(QtWidgets.QDoubleSpinBox):
+    """We are writing this method to override the spinbox in add_item so the increment/decrement is x10/x0.1"""
```
Comment (roomrys) [inline sleap/gui/dialogs/formbuilder.py]: We aren't concerned with what WE are doing, but rather what the class is:
```suggestion
    """A subclass of QDoubleSpinBox that increments and decrements by a factor of 10."""
```

> All class docstrings should start with a one-line summary that describes what the class instance represents.

Ref: [google-style docstrings](https://google.github.io/styleguide/pyguide.html#384-classes)

File: sleap/gui/dialogs/formbuilder.py [hunk 16]
```diff
@@ -458,14 +465,16 @@ def add_item(self, item: Dict[Text, Any]):
...
             return
 
-        # double: show spinbox (number w/ up/down controls)
-        elif item["type"] == "double":
-            
-            #If the label of the item is Loss Weight then we will set the min and max to 0 and 1000
-            if item["label"] == "Loss Weight":
+        elif (
+            item["type"] == "double"
+        ):  # double: show spinbox (number w/ up/down controls)
```
Comment (roomrys) [inline sleap/gui/dialogs/formbuilder.py]: I liked these better as block comments. Why? 
1. They're easier to see/read
2. The comment explains the block of code to follow not just the single line

```suggestion
        # double: show spinbox (number w/ up/down controls)
        elif item["type"] == "double":
```

File: sleap/gui/dialogs/formbuilder.py [hunk 17]
```diff
@@ -475,14 +484,12 @@ def add_item(self, item: Dict[Text, Any]):
...
             field.setRange(min, max)
-                
 
             field.setValue(item["default"])
 
             field.valueChanged.connect(lambda: self.valueChanged.emit())
 
-        # int: show spinbox (number w/ up/down controls)
-        elif item["type"] == "int":
+        elif item["type"] == "int":  # int: show spinbox (number w/ up/down controls)
```
Comment (roomrys) [inline sleap/gui/dialogs/formbuilder.py:487]: I liked these better as block comments. Why? 
1. They're easier to see/read
2. The comment explains the block of code to follow not just the single line

```suggestion
        # int: show spinbox (number w/ up/down controls)
        elif item["type"] == "int":
```

File: sleap/gui/dialogs/formbuilder.py [hunk 18]
```diff
@@ -458,14 +465,16 @@ def add_item(self, item: Dict[Text, Any]):
...
-            
-            #If the label of the item is Loss Weight then we will set the min and max to 0 and 1000
-            if item["label"] == "Loss Weight":
+        elif (
+            item["type"] == "double"
+        ):  # double: show spinbox (number w/ up/down controls)
+
+            if (
+                item["label"] == "Loss Weight"
+            ):  # Set min/max to 0-1000 if label is "Loss Weight"
```
Comment (roomrys) [inline sleap/gui/dialogs/formbuilder.py]: This bit of code is pretty self explanatory...
```suggestion
            if item["label"] == "Loss Weight": 
```
 if you really want to keep the comment make it a block comment. Why? The line you put the comment on isn't really what that line does is it? It's what happens in the following block. (Also, side note, the biggest thing is setting the ExponentialSPinBox, not changing the min value.)

File: tests/gui/test_formbuilder.py [hunk 19]
```diff
@@ -132,3 +140,93 @@ def test_string_list_widget(qtbot):
...
+    widget = formbuilder.ExponentialSpinBox()
+    qtbot.addWidget(widget)
+
+    widget.setValue(1.0)
+    print(widget.value())
+    assert math.isclose(widget.value(), 1.0, rel_tol=1e-3)
+
+    widget.stepBy(1)
```
Comment (roomrys) [inline tests/gui/test_formbuilder.py:154]: A few comments on what you are testing (in this test and others) would be nice, i.e.
```python
# Test that stepping by positive numbers multiplies value by 10
```

File: tests/gui/test_formbuilder.py [hunk 20]
```diff
@@ -132,3 +140,93 @@ def test_string_list_widget(qtbot):
...
+    print(widget.value())
+    assert math.isclose(widget.value(), 10.0, rel_tol=1e-3)
+
+    widget.stepBy(1)
```
Comment (roomrys) [inline tests/gui/test_formbuilder.py:157]: Also test other positive/negative values with magnitude not equal to 1.

File: tests/gui/test_formbuilder.py [hunk 21]
```diff
@@ -1,6 +1,7 @@
 import yaml
 
 from sleap.gui.dialogs import formbuilder
+import math
```
Comment (roomrys) [inline tests/gui/test_formbuilder.py:4]: Do we really need `math` here?
```suggestion
```

**Review by roomrys — CHANGES_REQUESTED** [review]
The tests are looking quite thorough!

Most pressing things:
1. Revert changes to .gitignore
2. Remove all print statements in tests (those won't help our automated testing)

Others:
1. Comments on comments and docstrings and linting
